builtin/logical/pki/acme_challenge_engine.go

```diff
@@ -305,22 +307,49 @@ func (ace *ACMEChallengeEngine) AcceptChallenge(sc *storageContext, account stri
 	return nil
 }
 
-func (ace *ACMEChallengeEngine) VerifyChallenge(runnerSc *storageContext, id string, finished chan bool, config *acmeConfigEntry) {
-	sc, _ /* cancel func */ := runnerSc.WithFreshTimeout(MaxChallengeTimeout)
+func (ace *ACMEChallengeEngine) VerifyChallenge(runnerSc *storageContext, id string, retries int, finished chan bool, config *acmeConfigEntry) {
+	sc, cancel := runnerSc.WithFreshTimeout(MaxChallengeTimeout)
+	defer cancel()
 	runnerSc.Backend.Logger().Debug("Starting verification of challenge", "id", id)
 
 	if retry, retryAfter, err := ace._verifyChallenge(sc, id, config); err != nil {
+		if errors.Is(err, logical.ErrReadOnly) {
+			// If we get a storage read only error we should bail out as we
+			// are no longer running on the active node as all ACME writes
+			// are cluster local
+			finished <- true
+			sc.Backend.Logger().Warn("ACME validation thread stopping as an ErrReadOnly error was encountered")
+			return
+		}
+
 		// Because verification of this challenge failed, we need to retry
 		// it in the future. Log the error and re-add the item to the queue
 		// to try again later.
 		sc.Backend.Logger().Error(fmt.Sprintf("ACME validation failed for %v: %v", id, err))
 
 		if retry {
+			retries++
+
+			// This additional retry logic is here in case we start failing to decode entries within the
+			// data store, we would re-attempt them forever and never clean them up. We want to leverage
+			// the retry counts within the authorization/challenge objects first which is why we give ourselves
+			// twice the attempts here.
+			if retries > MaxRetryAttempts*2 {
```

Comment:
I am not super familiar with the internals of acme challenge engine, so please bear with me with my questions. What is the correlation between MaxRetryAttempt and retry count within the authorization/challenge object? Would this doubling here cause confusion in the future as in it may cause two different behaviours for a single thing?

Reply:
> I am not super familiar with the internals of acme challenge engine, so please bear with me with my questions.

No worries, I appreciate the feedback!

> What is the correlation between MaxRetryAttempt and retry count within the authorization/challenge object?

There isn't really a correlation, but you are right that it may lead to confusion. Basically this new one I'm just trying to track are we looping continuously on a challenge that we can't seem to decode/process out of the normal ACME processing pipeline. Kinda a catch 22 in if we can't decode/read the event from storage how are we supposed to know how many times we've attempted it... 

I realized if we don't do this and for some reason these start to increase we may get stuck continuously processing these and nothing new incoming.  

> Would this doubling here cause confusion in the future as in it may cause two different behaviours for a single thing?

Maybe but again this is here as a stop gap, "should" never fire technically. 

One thing to keep in mind with ACME challenges/orders is, it isn't the end of the world to fail/lose them, not great mind you, but all the clients I've seen will/can just restart the process again with minimal effort.

Reply:
This is good context! Thanks!